darshan-runtime/lib/darshan.h

```diff
@@ -211,12 +299,66 @@ void *darshan_core_register_record(
 char *darshan_core_lookup_record_name(
     darshan_record_id rec_id);
 
+/* darshan_core_disabled_instrumentation
+ *
+ * Returns true (1) if Darshan has currently disabled instrumentation,
+ * false (0) otherwise. If instrumentation is disabled, modules should
+ * no longer update any file records as part of the intercepted function
+ * wrappers.
+ */
+static inline int darshan_core_disabled_instrumentation(void)
+{
+    int ret;
+
+    __DARSHAN_CORE_LOCK();
+    if(__darshan_core)
+        ret = 0;
+    else
+        ret = 1;
+    __DARSHAN_CORE_UNLOCK();
+
+    return(ret);
+}
+
+/* retrieve absolute wtime */
+static inline double darshan_core_wtime_absolute(void)
+{
+#ifdef __DARSHAN_RDTSCP_FREQUENCY
+    /* user configured darshan-runtime explicitly to use rtdscp for timing */
+    unsigned flag;
+    unsigned long long ts;
+
+    ts = __rdtscp(&flag);
+    return((double)ts/(double)__DARSHAN_RDTSCP_FREQUENCY);
+#else
+    /* normal path */
+    struct timespec tp;
+    /* some notes on what function to use to retrieve time as of 2021-05:
+     * - clock_gettime() is faster than MPI_Wtime() across platforms
+     * - clock_gettime() is at least competitive with gettimeofday()
+     * - on some platforms, the _COARSE variants may provide sufficient
+     *   resolution with a slight performance improvement, but there are a
+     *   couple of problems:
+     *   - the _COARSE variants are not implemented with vDSO on POWER
+     *     platforms
+     *   - it is not well defined how much precision will be sacrificed
+     */
+    clock_gettime(CLOCK_REALTIME, &tp);
+    return(((double)tp.tv_sec) + 1.0e-9 * ((double)tp.tv_nsec));
+#endif
+}
+
 /* darshan_core_wtime()
  *
  * Returns the elapsed time relative to (roughly) the start of
  * the application.
  */
-double darshan_core_wtime(void);
+static inline double darshan_core_wtime(void)
+{
+    double wtime_offset = __darshan_core ? __darshan_core->wtime_offset : 0;
```

Comment:
I think we actually need a lock around this, as in `darshan_core_disabled_insrumentation()`?

Reply:
Yeah, you are right.  I hate to add a lock there for a function called that frequently, though.

Even though it doesn't look as nice, it might be better to store the time offset in a global variable instead of within the __darshan_core pointer structure, so that it is always accessible.

I'll address this and update the PR.

Reply:
That make sense. The net result would be the same in any case, if the global variable `wtime_offset` is initialized to 0 (and I guess set to 0 at shutdown time, too).

Reply:
I took a crack at this via c3cb160.

`wtime_offset` is now a global. It is statically initialized to 0 and set at the same time `__darshan_core` is set in `darshan_core_initialize()`. Ultimately, this just means that if someone calls `darshan_core_wtime()` before Darshan has fully initialized, it will return absolute timestamps (rather than timestamps relative of job start). Otherwise, Darshan has already determined the job `wtime_offset`, and can return timestamps relative to that. That all sounds perfectly reasonable, in that callers shouldn't expect sensible results out of `darshan_core_wtime()` if they haven't ensured Darshan is actually initialized.